proposals/BVAL-214.md

```diff
@@ -0,0 +1,51 @@
+---
+title: Validate a list of changes without applying them to the actual object
+layout: default
+author: Gunnar Morling
+comments: true
+---
+
+# #{page.title}
+
+[Link to JIRA ticket][jira]  
+[Related JIRA](https://hibernate.atlassian.net/browse/BVAL-214)  
+
+## Problem
+
+When validating user input from a UI which is bound to the data model it is desirable to do the validation *before* propagating the values into the model.
+This prevents the model from being tainted with invalid values.
+For single properties there is `Validator#validateValue()` for this purpose, but an equivalent solution for class-level constraints is lacking today.
+
+## Proposition
+
+Provide a new method `<T> Set<ConstraintViolation<T>> Validator#validateValues(T bean, Map<String, Object> values, Class<?>... groups)`.
+This takes a bean instance and a map of property values.
+These values will be applied to a *clone* of the given object which then is validated.
+
+### Pros
+
+* The solution is transparent to class-level constraints, they'd see the validated bean as if it had the given property values
+* No need to pass in values through the map which haven't changed in the bean
+### Cons
+
+* Requires validated bean types to be clonable; That's a nasty requirement, esp. when it comes to JPA entities with lazy props etc.
+
+The requirement for clonability could possibly be mitigated by introducing a cloning SPI. By default, BV providers would resort to expecting beans to implement `Cloneable`. But alternative implementations could be based on less intrusive cloning approaches such as resorting to copy constructors or libraries such as https://github.com/kostaskougios/cloning
```

Comment:
I am very skeptical about the cloning part. Making bean `Clonable` is one thing, but you really would need deep clones. I think this is just too easy to get wrong (from the user side).

I guess using a library like _cloning_ might be a solution, but deep clones also come at a cost.